lib/plugins/aws/package/compile/events/stream/index.js

```diff
@@ -11,40 +11,93 @@ class AwsCompileStreamEvents {
+              },
+              {
+                type: 'object',
+                properties: {
+                  arn: { $ref: '#/definitions/awsCfFunction' },
+                  type: { enum: ['sns', 'sqs'] },
+                },
+                additionalProperties: false,
+                required: ['arn', 'type'],
+              },
+            ],
+          },
+        },
+        additionalProperties: false,
+        required: ['onFailure'],
+      },
+    });
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'stream', {
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      anyOf: [
+        { $ref: '#/definitions/awsArnString' },
+        {
+          type: 'object',
+          properties: {
+            arn: { $ref: '#/definitions/awsArnString' },
+            type: { enum: ['dynamodb', 'kinesis'] },
+            batchSize: { type: 'integer', minimum: 1, maximum: 10000 },
+            parallelizationFactor: { type: 'integer', minimum: 1, maximum: 10 },
+            startingPosition: { enum: ['LATEST', 'TRIM_HORIZON'] },
+            enabled: { type: 'boolean' },
+            consumer: { oneOf: [{ const: true }, { $ref: '#/definitions/awsArn' }] },
+            batchWindow: { type: 'integer', minimum: 0, maximum: 300 },
+            maximumRetryAttempts: { type: 'integer', minimum: -1, maximum: 10000 },
+            bisectBatchOnFunctionError: { type: 'boolean' },
+            maximumRecordAgeInSeconds: {
+              oneOf: [
+                { type: 'integer', minimum: -1, maximum: -1 },
+                { type: 'integer', minimum: 60, maximum: 604800 },
+              ],
+            },
+            destinations: { $ref: '#/definitions/streamDestinations' },
+          },
+          additionalProperties: false,
+          required: ['arn'],
+        },
+        {
+          type: 'object',
+          properties: {
+            arn: { $ref: '#/definitions/awsCfFunction' },
+            type: { enum: ['dynamodb', 'kinesis'] },
+            batchSize: { type: 'integer', minimum: 1, maximum: 10000 },
+            parallelizationFactor: { type: 'integer', minimum: 1, maximum: 10 },
+            startingPosition: { enum: ['LATEST', 'TRIM_HORIZON'] },
+            enabled: { type: 'boolean' },
+            consumer: { oneOf: [{ const: true }, { $ref: '#/definitions/awsArn' }] },
+            batchWindow: { type: 'integer', minimum: 0, maximum: 300 },
+            maximumRetryAttempts: { type: 'integer', minimum: -1, maximum: 10000 },
+            bisectBatchOnFunctionError: { type: 'boolean' },
+            maximumRecordAgeInSeconds: {
+              oneOf: [
+                { type: 'integer', minimum: -1, maximum: -1 },
+                { type: 'integer', minimum: 60, maximum: 604800 },
+              ],
+            },
+            destinations: { $ref: '#/definitions/streamDestinations' },
+          },
+          additionalProperties: false,
+          required: ['arn', 'type'],
+        },
+      ],
```

Comment:
Going that route is problematic for error reporting.

As e.g. if one makes error in `batchSize`, we will get error for both alternatives, and error normalizer, not being able to decide which alternative was chosen to be followed by user, will expose the error as generic `functions[].events[].sns: unsupported configuration format` which doesn't give a hint of what the real error is.

I think this can be cleanly solved solved by using `additionalProperties` as a container to define two options for `arn` an `type`, and keep others defined as it was in previous version

It also doesn't feel good, that we repeat definitions for most of properties, seems error-prone.

Reply:
Hi @medikoo , not sur what you mean by using `additionalProperties`. Could you provide an exemple ?

In the meantime, I found a solution detailed in https://github.com/json-schema/json-schema/issues/158 showing how to implement validation based on property value (the correct way of JSON schema is schema dependency based on property existence with `dependencies` keyword).

The resulting error message when specifying a CF intrinsic function without `type` is `Configuration error at 'functions.toto.events[0].stream.destinations.onFailure.arn': should be string` which seems correct. WDYT ?

Reply:
> Hi @medikoo , not sur what you mean by using additionalProperties. Could you provide an exemple ?

Sorry, that was an invalid suggestion, via `additionalProperties` you can enforce schema for _other_ properties, but obviously not _one_ by _one_, but for all generally.

> which seems correct. WDYT ?

Looks great to me!